### ISSUE TITLE
[Bug]: range_search with radius  lost valid data

### ISSUE THREAD

**Issue by wuweicheng39** [milvus-io/milvus#31434]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version:
  2.3.12
- Deployment mode(standalone or cluster):
  docker-compose standalone
- MQ type(rocksmq, pulsar or kafka):  
  -  
- SDK version(e.g. pymilvus v2.0.0rc2):
  pymilvus 2.3.7
- OS(Ubuntu or CentOS): 
  centos
- CPU/Memory: 
- GPU: 
- Others:

entity_num :  2000w
dim_size: 1024
index_type: ivf_sq8
topk = 10
expr = None

search_params = {
    'metric_type': 'IP',
    'params': {
        'nprobe':10,
        'radius': 0.4,
#        'range_filter' : 1.2,
    }
}
```

### Current Behavior

I try the range_search with radius，but get the unexpceted result，lost some valid data, comparing without range_search.

image: range_search with radius=0.4
<img width="478" alt="截屏2024-03-20 14 20 14" src="https://github.com/milvus-io/milvus/assets/94667927/f4009e04-6630-4131-876e-9d9986488e8d">

### Expected Behavior

image: without range_search

<img width="442" alt="截屏2024-03-20 14 22 01" src="https://github.com/milvus-io/milvus/assets/94667927/6a953388-c441-4e74-9d5a-99ad0ccd2aed">

**Comment by wuweicheng39** [comment]
search_params = {
    'metric_type': 'IP',
    'params': {
        'nprobe':10,
        # search for vectors with a distance smaller than 1.0
#        'radius': 0.4,
#        # filter out vectors with a distance smaller than or equal to 0.8
#        'range_filter' : 1.2,
    }
}

**Comment by xiaofan-luan** [comment]
Range search might not be exactaly same as topk search.
@wuweicheng39 how many is different

**Comment by wuweicheng39** [comment]
you can see two images above

**Comment by xiaofan-luan** [comment]
like for top100, how many is in topk but not rangesearch?
Miss 10 result might be reasonable

**Comment by wuweicheng39** [comment]
topk = 10

**Comment by xiaofan-luan** [comment]
> topk = 10

If 10 results are not same, this might be simply a search accuracy difference.

@liliu-z  @cydrain 
how to increase range search accuracy? Any knob to tune?

**Comment by liliu-z** [comment]
Looks like we did range search and get top10 from it. Comparing with normal KNN search without range search, it lost 2 valid results.
This is caused by different search algorithm behind these two different kinds of searches. From Milvus 2.4 we can tune the accuracy by modify param `max_empty_result_buckets` https://milvus.io/docs/index.md#IVF_SQ8.

**Comment by xiaofan-luan** [comment]
> Looks like we did range search and get top10 from it. Comparing with normal KNN search without range search, it lost 2 valid results. This is caused by different search algorithm behind these two different kinds of searches. From Milvus 2.4 we can tune the accuracy by modify param `max_empty_result_buckets` https://milvus.io/docs/index.md#IVF_SQ8.

the default value seems to be too small. Let's get at least 2.

**Comment by xiaofan-luan** [comment]
and the introduction of the parameter is very confusing, range should be [1, nlist] something like that.

and we should give out better explain.  "Maximum number of empty buckets that has been returned." -> Maximum number of empty buckets searched before we return

**Comment by wuweicheng39** [comment]
how to resolve the problem???

**Comment by xiaofan-luan** [comment]
change max_empty_result_buckets to larger than value may work

**Comment by cydrain** [comment]
> how to resolve the problem???

Hi @wuweicheng39 ,

**Comment by cydrain** [comment]
> change max_empty_result_buckets to larger than value may work

this range search parameter is introduced in Milvus 2.4, it's not available in Milvus 2.3.12

**Comment by wuweicheng39** [comment]
> > change max_empty_result_buckets to larger than value may work
> 
> this range search parameter is introduced in Milvus 2.4, it's not available in Milvus 2.3.12

thanks, in Milvu2.3.12, I tried to set  max_empty_result_buckets=1024, but it also did not work.  later I will try Milvus 2.4.

**Comment by cydrain** [comment]
> how to resolve the problem???

Except 'radius' and 'range_filter', there is no other parameters to tune the range search behavior for IVF-serial indices in Milvus 2.3.x.
I suggest to try Milvus 2.4.0, or you can try HNSW range search which can improve its recall by increasing 'ef'.

**Comment by cydrain** [comment]
> > > change max_empty_result_buckets to larger than value may work
> > 
> > 
> > this range search parameter is introduced in Milvus 2.4, it's not available in Milvus 2.3.12
> 
> thanks, in Milvu2.3.12, I tried to set max_empty_result_buckets=1024, but it also did not work. later I will try Milvus 2.4.

I can explain more here,

**Comment by yanliang567** [comment]
@wuweicheng39 any updates for milvus 2.4?
